server/src/main/java/keywhiz/service/resources/admin/SecretsResource.java

```diff
@@ -487,6 +490,135 @@ public Response deleteSecret(@Auth User user, @PathParam("secretId") LongParam s
     return Response.noContent().build();
   }
 
+  /**
+   * Finds all deleted secrets that have the queried name, which will be queried as "." + name + ".%"
+   *
+   * @param user        the admin user performing this operation
+   * @param name        Secret series deleted name
+   * @return a list of deleted secrets that have the name
+   * <p>
+   * responseMessage 200 Secret series information retrieved
+   * <p>
+   * responseMessage 404 Secret series not found
+   */
+  @Timed @ExceptionMetered
+  @GET
+  @Path("deleted/{name}")
+  @Produces(APPLICATION_JSON)
+  public List<SecretSeries> findDeletedSecretsByName(@Auth User user,
+      @PathParam("name") String name) {
+
+    logger.info("User '{}' finding deleted secrets with name '{}'.", user, name);
+
+    return secretDAOReadOnly.getSecretsWithDeletedName(name);
+  }
+
+  /**
+   * Rename Secret by ID to the given name
+   *
+   * @param user     the admin user performing this operation
+   * @param secretId the ID of the Secret to be deleted
+   * @return 200 if secret deleted, 404 if not found
+   * <p>
+   * description Deletes a single Secret if found. Used by Keywhiz CLI and the web ui.
+   * <p>
+   * responseMessage 200 Found and deleted Secret with given ID
+   * <p>
+   * responseMessage 404 Secret with given ID not Found
+   */
+  @Path("rename/{secretId}/{secretName}")
+  @Timed @ExceptionMetered
+  @POST
+  public Response renameSecret(@Auth User user, @PathParam("secretId") LongParam secretId,
+      @PathParam("secretName") String secretName) {
+    Optional<Secret> secret = secretController.getSecretByName(secretName);
+    if (secret.isPresent()) {
+      logger.info("User '{}' tried renaming a secret, but another secret with that name "
+              + "already exists (name={})", user, secretId.get());
+      throw new NotAllowedException("That name is already taken by another secret"); //use different error
+    }
+
+    logger.info("User '{}' renamed secret id={} to name='{}'", user, secretId,
+        secretName);
+
+    secretDAOReadWrite.renameSecretById(secretId.get(), secretName);
```

Comment:
This will throw a `DataAccessException` if the secret name already exists; we should catch it and throw an exception here as well.